### PR TITLE
fix(scheduler): keep active slot occupied when launch succeeds but bookkeeping fails

### LINKED ISSUE #4452: [bug] Scheduler can launch a second run after bookkeeping failure
## Before you start

- [x] I searched the existing issues and open pull requests; I did not find a duplicate.
- [x] Reproduced against current `main` (`3b77a7401b549fa6da4c8e1f8c2c0081d56e3d7a`).

## Problem summary

A transient bookkeeping failure after `_launch_run` succeeds can mark the scheduled-task row terminal and free its overlap slot while the launched run is still active.

## Affected area

Backend scheduler (`backend/app/scheduler/service.py`, `ScheduledTaskService.dispatch_task`).

## What happened?

The active-row constraint added in #4105 protects concurrent row creation, but there is still a post-launch window:

1. `dispatch_task` creates a `queued` task-run row.
2. `_launch_run` succeeds and returns a live `run_id`.
3. The subsequent `update_status(..., status=running)` raises, for example during a transient database failure.
4. The broad exception handler updates the same row to `failed`, which removes it from the partial unique index for active task runs.
5. A later dispatch sees no active task run and can launch a second run while the first one is still live.

The failure result also returns `run_id: None`, so the scheduler loses the identifier that could be used for reconciliation.

## Expected behavior

Once a run has been launched successfully, a later bookkeeping error must not release the active slot until that run is known to be terminal. The launched `run_id` should remain recoverable for reconciliation or cancellation.

## Steps to reproduce

A deterministic service-level reproduction is:

1. Use a task-run repository whose first `update_status` call after launch raises, while the failure-path status update succeeds.
2. Use a `launch_run` stub that records each launch and returns a live run id.
3. Dispatch the same scheduled task twice.

Observed result:

```text
first_outcome=failed
second_outcome=launched
launch_count=2
```

No production logs are required; the failure is reproducible by injecting the single database write failure at the exact post-launch transition.

## Impact

This can bypass the one-active-run-per-task invariant during a narrow database-failure window, causing duplicate scheduled work and losing the first run id from scheduler bookkeeping.

## Suggested acceptance criteria

- A failure after `_launch_run` succeeds cannot make the task eligible for another launch while that run is live.
- The launched `run_id` is retained for recovery, reconciliation, or cancellation.
- A regression test injects failure in the `queued -> running` write and verifies that a second dispatch does not launch another run.

This was first identified by `willem-bd` in the [review of #4105](https://github.com/bytedance/deer-flow/pull/4105#pullrequestreview-4684132500); that PR correctly treated it as a pre-existing, out-of-scope edge case.

### PR DESCRIPTION
**Pull request by YeyezhizzZ**
Fixes #4452

## Why

A transient database failure after `_launch_run` succeeds can mark the scheduled-task row terminal and free its overlap slot while the launched run is still active. This was first identified in the review of #4105 and filed as #4452. The previously attempted fix (#4473) was closed because the recovery branch diverged from the success path for cron tasks — hardcoding `status="running"` caused cron tasks to permanently stop firing, and the once-task status fell through to `"enabled"` instead of `"running"`.

This PR implements the reviewer-recommended approach: capture `_task_status` (computed from the success path's own `schedule_type` / `trigger` branches) before any bookkeeping write that can throw, then reuse it in the recovery branch so it stays in lockstep for all three cases.

## What changed

- `dispatch_task` now tracks whether `_launch_run` succeeded and, if so, what `task_status` the success path would have written. When a bookkeeping write fails after a successful launch, the recovery branch keeps the run row as `"running"` and reuses the captured `task_status` rather than re-deriving it (which previously diverged for once and cron tasks).
- New test file with four regression tests pinning the slot-occupancy invariant and the per-type status invariant (once → `"running"`, cron → `"enabled"`, launch-fail → slot released).

## Surface area

- [x] **Backend API** — scheduler dispatch path under `backend/app`
- [ ] **Frontend UI**
- [ ] **Agents / LangGraph**
- [ ] **Sandbox**
- [ ] **Skills**
- [ ] **Dependencies**
- [ ] **Default behavior change**
- [ ] **Docs / tests / CI only**

## Bug fix verification

- **Test path:** `backend/tests/test_scheduled_task_bookkeeping_failure.py`
- **Did it go red on `main`?** Yes — the three "launch succeeds, bookkeeping fails" tests fail on `main` because the existing except block unconditionally sets `status="failed"`, releasing the active slot and (for once/cron) setting the wrong parent status. The "launch fails" test passes on both.
- All 60 scheduler-related tests pass on this branch (56 existing + 4 new).

## Validation

```bash
cd backend && make lint  # All checks passed
cd backend && make test  # 60 passed (scheduler tests)
```

## AI assistance

**Tool(s) used:** Claude Code

**How you used it:** Explored the codebase to understand the scheduler dispatch flow and the `uq_scheduled_task_run_active` partial unique index. Reviewed the #4473 PR discussion and reviewer feedback to understand what went wrong in the previous attempt. The AI helped draft the implementation based on the reviewer-recommended approach (capturing `_task_status` from the success path's own conditional, not hardcoding or re-deriving it). Tests were structured to cover all three status branches plus the launch-fail case.

- [x] I've read and understand every line of this change and take responsibility for it — it's not unreviewed AI output.

### CODE REVIEW

File: backend/app/scheduler/service.py [hunk 1]
```diff
@@ -198,6 +206,46 @@ async def dispatch_task(
...
+                    next_run_at=next_at,
+                    last_run_at=now,
+                    last_run_id=_launch_run_result["run_id"],
+                    last_thread_id=_launch_run_result["thread_id"],
+                    last_error=str(exc),
+                    increment_run_count=True,
+                    protect_terminal=True,
+                )
+                return {
+                    "outcome": "failed",
```
Comment (willem-bd) [inline backend/app/scheduler/service.py]: Heads-up on the outcome semantics: this branch returns `"failed"` while a run is actually live in the external scheduler (`run_id` is populated). Every other `"failed"` return in `dispatch_task` (line 268, and the launch-failure test asserts `result["run_id"] is None`) implies "no run was launched". A caller that branches on `outcome == "failed"` to mean "nothing is running, safe to retry / nothing to cancel" will be wrong here.

For the scheduled poll path this is logged and ignored, so it's low-impact today. But if a manual trigger ever hits this branch, the router would surface `"failed"` to the user while a run is silently executing. Worth either a distinct outcome (e.g. `"launched_bookkeeping_failed"`) or a short code comment documenting the overload so future callers handle the non-None `run_id` correctly.

File: backend/app/scheduler/service.py [hunk 2]
```diff
@@ -198,6 +206,46 @@ async def dispatch_task(
...
+            # still running, and preserve the run_id for reconciliation.
+            if _launch_run_succeeded:
+                # Reuse the success-path task_status so the recovery branch
+                # stays in lockstep for all three cases:
+                #   once           → "running" (cancel_stuck_once_tasks can reconcile)
+                #   cron           → "enabled" (claim_due_tasks can reclaim)
+                #   manual-paused  → "paused"
+                # Fall back only if next_run_at itself raised before _task_status
+                # was assigned.
+                task_status = _task_status if _task_status is not None else self._task_status_for_failure(task, trigger=trigger)
```
Comment (willem-bd) [inline backend/app/scheduler/service.py:220]: Minor: the `_task_status is not None` fallback (and the comment above it) describe a path that's effectively unreachable. `_launch_run_succeeded` is set at line 153, and the only statements between `next_at = next_run_at(...)` (line 154) and `_task_status = ...` (line 165) are pure `if/elif/else` dict checks — no await point. So `_task_status` can only still be `None` here if `next_run_at` raised in the try block; but then the second `next_run_at(...)` call at line 200 raises again (it's a pure function of the same inputs) and propagates uncaught, so control never reaches this `if _launch_run_succeeded:` block.

The defensive fallback is harmless, but the comment overstates the case it handles — worth trimming so future readers don't reason about a phantom edge case (or, if you want to keep the guard, drop the explanatory comment).

File: backend/tests/test_scheduled_task_bookkeeping_failure.py [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+    result = await service.dispatch_task(dict(task), now=datetime.now(UTC), trigger="scheduled")
+
+    assert len(launched) == 1
+    assert launched[0]["metadata"]["scheduled_task_id"] == "task-bf-cron"
+    assert result["outcome"] == "failed"
+    assert result["run_id"] is not None
+    assert result["run_id"] == "run-1"
+
+    # After recovery the run row is "running", so has_active_runs returns True.
+    run_repo._active = True
```
Comment (willem-bd) [inline backend/tests/test_scheduled_task_bookkeeping_failure.py:152]: This assertion is slightly circular: line 147 sets `run_repo._active = True` and line 148 asserts `has_active_runs(...) is True`, which only confirms the fake returns what it was just assigned. It doesn't verify the actual production mechanism that keeps the slot occupied — namely that the recovery branch re-wrote the run row as `status="running"` with the live `run_id`.

Consider asserting against `run_repo.updated` directly, e.g.:

```python
last_update = run_repo.updated[-1][1]
assert last_update["status"] == "running"
assert last_update["run_id"] == "run-1"
```

That pins the recovery write itself and would catch a regression even if `has_active_runs` semantics change later. The same pattern applies to the once-task test below.

**Review by willem-bd — COMMENTED** [review]
Thanks for the clean fix — capturing `_task_status` from the success path's own `schedule_type` / `trigger` branches (rather than re-deriving or hardcoding) is the right call, and the per-status-type regression tests pin the invariant well. Three smaller follow-ups below, none of them blocking.

**Comment by YeyezhizzZ** [comment]
Thanks for the review @willem-bd! All three items addressed:

**1. Distinct outcome** ([\#discussion_r3670578612](https://github.com/bytedance/deer-flow/pull/4557#discussion_r3670578612)):
- Changed recovery-branch outcome from `"failed"` to `"launched_bookkeeping_failed"` — a run IS live here (`run_id` is non-None), unlike every other `"failed"` return. Added a comment documenting the semantic difference.

**2. Unreachable-fallback comment** ([\#discussion_r3670578618](https://github.com/bytedance/deer-flow/pull/4557#discussion_r3670578618)):
- Removed the misleading comment about `_task_status is not None` fallback. The defensive guard is kept as harmless, but the comment no longer describes a phantom edge case.

**3. Circular test assertions** ([\#discussion_r3670578620](https://github.com/bytedance/deer-flow/pull/4557#discussion_r3670578620)):
- Replaced `has_active_runs` circular assertions with direct `run_repo.updated` checks that pin the recovery-branch `status="running"` + `run_id` write. Applied to both `test_launch_succeeds_but_bookkeeping_fails_keeps_slot_occupied` and `test_once_task_bookkeeping_failure_keeps_status_running`.

**Validation:** All 64 scheduler-related tests pass.

Fixed in `28498a4`.

File: backend/app/scheduler/service.py (hunk 1 above)
Comment (YeyezhizzZ) [reply backend/app/scheduler/service.py]: Thanks for the review.

**Updated:**
- Changed outcome from `"failed"` to `"launched_bookkeeping_failed"` so callers can distinguish "no run launched" from "run launched, bookkeeping broken". Added a comment documenting the semantic difference.

Fixed in `28498a4`.

File: backend/app/scheduler/service.py (hunk 2 above)
Comment (YeyezhizzZ) [reply backend/app/scheduler/service.py:220]: Thanks — you're right, that path is unreachable. Removed the misleading comment; the defensive guard is kept as harmless.

File: backend/tests/test_scheduled_task_bookkeeping_failure.py (hunk 3 above)
Comment (YeyezhizzZ) [reply backend/tests/test_scheduled_task_bookkeeping_failure.py:152]: Good catch. Both affected tests now assert against `run_repo.updated` directly, pinning the recovery-branch `status="running"` + `run_id` write instead of the circular `has_active_runs` check. All 64 scheduler tests pass.

**Comment by WillemJiang** [comment]
The issue had been resolved by #4504
